bcipy/helpers/triggers.py

```diff
@@ -207,6 +214,110 @@ class FlushFrequency(Enum):
     END = "flush at end of session"
 
 
+def read_data(lines: Iterable[str]) -> List[Trigger]:
+    """Read raw trigger data from the given source.
+
+    Parameters
+    ----------
+        data - iterable object where each item is a str with data for a single
+            trigger.
+
+    Returns
+    -------
+        list of all Triggers in the data.
+    """
+    triggers = []
+    for i, line in enumerate(lines):
+        try:
+            trg = Trigger.from_list(line.split())
+            triggers.append(trg)
+        except (AssertionError, ValueError) as trg_error:
+            raise BciPyCoreException(
+                f'Error reading trigger from line {i+1}: {trg_error}'
+            ) from trg_error
+    return triggers
+
+
+def find_starting_offset(triggers: List[Trigger], device_type: str = None) -> Trigger:
+    """Given a list of raw trigger data, determine the starting offset for the
+    given device. The returned trigger has the timestamp of the first sample
+    recorded for the device.
+
+    If no device is provided the EEG offset will be used. If there are
+    no offset triggers in the given data a Trigger with offset of 0.0 will be
+    returned.
+
+    Parameters
+    ----------
+        triggers - list of trigger data; should include Triggers of
+          TriggerType.OFFSET
+        device_type - each device will generally have a different offset. This
+            parameter is used to determine which trigger to use. If not given
+            the EEG offset will be used by default. Ex. 'EYETRACKER'
+    Returns
+    -------
+        The Trigger for the first matching offset for the given device, or a
+            Trigger with offset of 0.0 if a matching offset was not found.
+    """
+    label = 'starting_offset'
+    if device_type and device_type != 'EEG':
+        label += f"_{device_type}"
+    for trg in triggers:
+        if trg.type == TriggerType.OFFSET and trg.label == label:
+            return trg
+    log.info(f"Offset not found (device_type: {device_type}); using 0.0")
+    return Trigger('starting_offset', TriggerType.OFFSET, 0.0)
+
+
+def read(path: str) -> List[Trigger]:
+    """Read all Triggers from the given text file.
+
+    Parameters
+    ----------
+        path - trigger (.txt) file to read
+    Returns
+    -------
+        triggers
+    """
+    if not path.endswith('.txt') or not os.path.exists(path):

```

Comment:
this seems like a more general-purpose helper that could be put in load.py

Reply:
That's an interesting suggestion. I like the idea of making `read` more generic, with other modules providing a function specifying what to do with the data. Maybe something like:

    def read(path: str, read_data: Callable[Iterable[str], Any]) -> Any:
        """Read data from the given file.

        Parameters
        ----------
            path - file to read
            read_data - function used to process each line of the file
        """
        if not os.path.exists(path):
            raise FileNotFoundError(f'File not found at [{path}].')
        with open(path, encoding=DEFAULT_ENCODING) as raw_txt:
            return read_data(raw_txt)


In our current module structure this approach might be problematic with cyclic dependency loading. We also lose out on some specificity in the messaging unless we add more parameters, in which case we lose the advantage of abstracting this out. I'll have to give this some more thought.

Reply:
Yeah, thats exactly what I had in mind. For now, though, could just have a helper to read .txt files in load.py.